packages/eslint-plugin/src/rules/key-spacing.ts

```diff
@@ -0,0 +1,404 @@
+            },
+          });
+        }
+
+        if (align === 'colon') {
+          checkAfterColon(node, nAfterColon, mode);
+        } else {
+          checkBeforeColon(node, nBeforeColon, mode);
+        }
+      }
+    }
+
+    function checkIndividualNode(
+      node: TSESTree.Node,
+      { singleLine }: { singleLine: boolean },
+    ): void {
+      const beforeColon =
+        (singleLine
+          ? options.singleLine
+            ? options.singleLine.beforeColon
+            : options.beforeColon
+          : options.multiLine
+          ? options.multiLine.beforeColon
+          : options.beforeColon) ?? false;
+      const nBeforeColon = beforeColon ? 1 : 0;
+      const afterColon =
+        (singleLine
+          ? options.singleLine
+            ? options.singleLine.afterColon
+            : options.afterColon
+          : options.multiLine
+          ? options.multiLine.afterColon
+          : options.afterColon) ?? true;
+      const nAfterColon = afterColon ? 1 : 0;
+      const mode =
+        (singleLine
+          ? options.singleLine
+            ? options.singleLine.mode
+            : options.mode
+          : options.multiLine
+          ? options.multiLine.mode
+          : options.mode) ?? 'strict';
+
+      if (isKeyTypeNode(node)) {
+        checkBeforeColon(node, nBeforeColon, mode);
+        checkAfterColon(node, nAfterColon, mode);
+      }
+    }
+
+    function validateBody(
+      body:
+        | TSESTree.TSTypeLiteral
+        | TSESTree.TSInterfaceBody
+        | TSESTree.ClassBody,
+    ): void {
+      const isSingleLine = body.loc.start.line === body.loc.end.line;
+
+      const members = 'members' in body ? body.members : body.body;
+
+      let alignGroups: TSESTree.Node[][] = [];
+      let unalignedElements: TSESTree.Node[] = [];
+
+      if (options.align || options.multiLine?.align) {
+        let currentAlignGroup: TSESTree.Node[] = [];
+        alignGroups.push(currentAlignGroup);
+
+        let prevNode: TSESTree.Node | undefined = undefined;
+
+        for (const node of members) {
+          let prevAlignedNode = currentAlignGroup.at(-1);
+          if (prevAlignedNode !== prevNode) {
+            prevAlignedNode = undefined;
+          }
+
+          if (prevAlignedNode && continuesAlignGroup(prevAlignedNode, node)) {
+            currentAlignGroup.push(node);
+          } else if (prevNode?.loc.start.line === node.loc.start.line) {
+            if (prevAlignedNode /* prevNode === prevAlignedNode */) {
+              unalignedElements.push(currentAlignGroup.pop()!);
+            }
+            unalignedElements.push(node);
+          } else {
+            currentAlignGroup = [node];
+            alignGroups.push(currentAlignGroup);
+          }
+
+          prevNode = node;
+        }
+
+        unalignedElements = unalignedElements.concat(
+          alignGroups
+            .filter(group => group.length === 1)
+            .flatMap(group => group),
+        );
```

Comment:
[Non-blocking] Just throwing it out there, you could do this to save a line...

```suggestion
        unalignedElements = unalignedElements.concat(
          ...alignGroups.filter(group => group.length === 1),
        );
```

If you have even the _slightest_ preference for the way it is now, feel free to ignore me. Just thought I'd mention the slightly smaller alternative! 😄